### ISSUE TITLE
Questions about downsample

### ISSUE THREAD

**Issue by Vegetebird** [fabro66/GAST-Net-3DPoseEstimation#26]
Hi~

The paper trained the network with downsample as 5, but VideoPose downsample is 1. When I train the network with downsample as 1, the performance is wrose than VideoPose.

So I want to know what's the performance you trained with downsample as 1?

**Comment by fabro66** [comment]
Hi~
The performance is worse than VideoPose with downsample as 1. This is because human3.6M dataset is captured at 50Hz, and there is a lot of redundant information between adjacent frames, which is not conducive to the extraction of spatial features.  If we set the downsample to 5, it could help learn the diversity of the 3D pose spatial structure. In our results, we infer customed videos in the wild without downsampling videos. 

There are several ways to improve the performance of GAST-Net：
1) introduce a multi-stride temporal convolution at the input layer (I did it, reduce errors about 1.0)
2) add a temporal attention mechanism different from the [paper](https://ieeexplore.ieee.org/document/9156272) (I did it, reduce errors about 0.5mm)
3) add a 2D poses correction layer (to do)
4) introduce GAN training method (doing)
5) Explore the role of global attention in half-body reconstruction (doing)

Another problem is that SemGCN will increase a lot of training time. Here, I provide you with another lightweight GCNs for you, which basically without reducing the accuracy but can speed up training.

```
from __future__ import absolute_import, division

import math
import torch
import torch.nn as nn
import torch.nn.functional as F
import numpy as np


class SemGraphConv(nn.Module):
    """
    Semantic graph convolution layer
    """

    def __init__(self, in_features, out_features, adj, bias=True):
        super(SemGraphConv, self).__init__()
        self.in_features = in_features
        self.out_features = out_features

        self.W = nn.Parameter(torch.zeros(size=(2, in_features, out_features), dtype=torch.float))
        nn.init.xavier_uniform_(self.W.data, gain=1.414)

        self.adj = adj
        self.m = (self.adj > 0)
        self.e = nn.Parameter(torch.zeros(1, len(self.m.nonzero()), dtype=torch.float))
        nn.init.constant_(self.e.data, 1)

        if bias:
            self.bias = nn.Parameter(torch.zeros(out_features, dtype=torch.float))
            stdv = 1. / math.sqrt(self.W.size(2))
            self.bias.data.uniform_(-stdv, stdv)
        else:
            self.register_parameter('bias', None)

    def forward(self, input):
        # X: (B, T, K, C)

        h0 = torch.matmul(input, self.W[0])
        h1 = torch.matmul(input, self.W[1])

        adj = -9e15 * torch.ones_like(self.adj).to(input.device)
        adj[self.m] = self.e
        adj = F.softmax(adj, dim=1)

        M = torch.eye(adj.size(0), dtype=torch.float).to(input.device)

        output = torch.matmul(adj * M, h0) + torch.matmul(adj * (1 - M), h1)

        if self.bias is not None:
            return output + self.bias.view(1, 1, -1)
        else:
            return output

    def __repr__(self):
        return self.__class__.__name__ + ' (' + str(self.in_features) + ' -> ' + str(self.out_features) + ')'


class LocalGraph(nn.Module):
    def __init__(self, adj, input_dim, output_dim, dropout=None):
        super(LocalGraph, self).__init__()

        num_joints = adj.shape[0]

        # Human3.6M
        if num_joints == 17:
            store_2 = [3, 6, 10, 13, 16]
            joints_left = [4, 5, 6, 11, 12, 13]
            joints_right = [1, 2, 3, 14, 15, 16]

        # Human3.6M detected from Stacked Hourglass
        elif num_joints == 16:
            store_2 = [3, 6, 9, 12, 15]
            joints_left = [4, 5, 6, 10, 11, 12]
            joints_right = [1, 2, 3, 13, 14, 15]

        # HumanEva
        elif num_joints == 15:
            store_2 = [4, 7, 10, 13]
            joints_left = [2, 3, 4, 8, 9, 10]
            joints_right = [5, 6, 7, 11, 12, 13]

        else:
            print('num_joints: %d' % num_joints)
            raise KeyError("The dimension of adj matrix is wrong!")

        adj_sym = torch.zeros_like(adj)
        for i in range(num_joints):
            for j in range(num_joints):
                if i == j:
                    adj_sym[i][j] = 1
                if i in joints_left:
                    index = joints_left.index(i)
                    adj_sym[i][joints_right[index]] = 1.0
                if i in joints_right:
                    index = joints_right.index(i)
                    adj_sym[i][joints_left[index]] = 1.0

        adj_1 = adj.matrix_power(1)
        # store_2 = [3, 6, 10, 13, 16]
        for i in np.arange(num_joints):
            if i in store_2:
                adj_1[i] = 0

        adj_2 = adj.matrix_power(2)
        # store_2 = [3, 6, 10, 13, 16]
        for i in np.arange(num_joints):
            if i not in store_2:
                adj_2[i] = 0

        adj_con = adj_1 + adj_2

        self.gcn_sym = SemGraphConv(input_dim, output_dim, adj_sym)
        self.bn_1 = nn.BatchNorm2d(output_dim, momentum=0.1)
        self.gcn_con = SemGraphConv(input_dim, output_dim, adj_con)
        self.bn_2 = nn.BatchNorm2d(output_dim, momentum=0.1)
        self.relu = nn.ReLU()

        self.cat_conv = nn.Conv2d(2 * output_dim, output_dim, 1, bias=False)
        self.cat_bn = nn.BatchNorm2d(output_dim, momentum=0.1)

        if dropout is not None:
            self.dropout = nn.Dropout2d(dropout)
        else:
            self.dropout = None

    def forward(self, input):
        # x: (B, T, K, C)
        x = self.gcn_sym(input)
        y = self.gcn_con(input)

        # x: (B, T, K, C) --> (B, C, T, K)
        x = x.permute(0, 3, 1, 2)
        y = y.permute(0, 3, 1, 2)

        x = self.relu(self.bn_1(x))
        y = self.relu(self.bn_2(y))

        output = torch.cat((x, y), dim=1)
        output = self.cat_bn(self.cat_conv(output))

        if self.dropout is not None:
            output = self.dropout(self.relu(output))
        else:
            output = self.relu(output)
        output = output.permute(0, 2, 3, 1)

        return output


```

**Comment by Vegetebird** [comment]
你好，感谢你详细的回复。我也是做3D姿态估计的，可以加个微信以后多多交流吗？我微信是：Vegetaberd




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "fabro66/GAST-Net-3DPoseEstimation"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年12月5日(星期六) 下午4:34
收件人:&nbsp;"fabro66/GAST-Net-3DPoseEstimation"<GAST-Net-3DPoseEstimation@noreply.github.com&gt;;
抄送:&nbsp;"wenhaoli"<781256005@qq.com&gt;;"Author"<author@noreply.github.com&gt;;
主题:&nbsp;Re: [fabro66/GAST-Net-3DPoseEstimation] Questions about downsample (#26)





 
Hi~
 The performance is worse than VideoPose with downsample as 1. This is because human3.6M dataset is captured at 50Hz, and there is a lot of redundant information between adjacent frames, which is not conducive to the extraction of spatial features.  If we set the downsample to 5, it could help learn the diversity of the 3D pose spatial structure. In our results, we infer customed videos in the wild without downsampling videos.
 
There are several ways to improve the performance of GAST-Net：
  
introduce a multi-stride temporal convolution at the input layer (I did it, reduce errors about 1.0)
 
add a temporal attention mechanism different from the paper (I did it, reduce errors about 0.5mm)
 
add a 2D poses correction layer (to do)
 
introduce GAN training method (doing)
 
Explore the role of global attention in half-body reconstruction (doing)
  
Another problem is that SemGCN will increase a lot of training time. Here, I provide you with another lightweight GCNs for you, which basically without reducing the accuracy but can speed up training.
 from __future__ import absolute_import, division import math import torch import torch.nn as nn import torch.nn.functional as F import numpy as np class SemGraphConv(nn.Module):     """     Semantic graph convolution layer     """     def __init__(self, in_features, out_features, adj, bias=True):         super(SemGraphConv, self).__init__()         self.in_features = in_features         self.out_features = out_features         self.W = nn.Parameter(torch.zeros(size=(2, in_features, out_features), dtype=torch.float))         nn.init.xavier_uniform_(self.W.data, gain=1.414)         self.adj = adj         self.m = (self.adj &gt; 0)         self.e = nn.Parameter(torch.zeros(1, len(self.m.nonzero()), dtype=torch.float))         nn.init.constant_(self.e.data, 1)         if bias:             self.bias = nn.Parameter(torch.zeros(out_features, dtype=torch.float))             stdv = 1. / math.sqrt(self.W.size(2))             self.bias.data.uniform_(-stdv, stdv)         else:             self.register_parameter('bias', None)     def forward(self, input):         # X: (B, T, K, C)         h0 = torch.matmul(input, self.W[0])         h1 = torch.matmul(input, self.W[1])         adj = -9e15 * torch.ones_like(self.adj).to(input.device)         adj[self.m] = self.e         adj = F.softmax(adj, dim=1)         M = torch.eye(adj.size(0), dtype=torch.float).to(input.device)         output = torch.matmul(adj * M, h0) + torch.matmul(adj * (1 - M), h1)         if self.bias is not None:             return output + self.bias.view(1, 1, -1)         else:             return output     def __repr__(self):         return self.__class__.__name__ + ' (' + str(self.in_features) + ' -&gt; ' + str(self.out_features) + ')' class LocalGraph(nn.Module):     def __init__(self, adj, input_dim, output_dim, dropout=None):         super(LocalGraph, self).__init__()         num_joints = adj.shape[0]         # Human3.6M         if num_joints == 17:             store_2 = [3, 6, 10, 13, 16]             joints_left = [4, 5, 6, 11, 12, 13]             joints_right = [1, 2, 3, 14, 15, 16]         # Human3.6M detected from Stacked Hourglass         elif num_joints == 16:             store_2 = [3, 6, 9, 12, 15]             joints_left = [4, 5, 6, 10, 11, 12]             joints_right = [1, 2, 3, 13, 14, 15]         # HumanEva         elif num_joints == 15:             store_2 = [4, 7, 10, 13]             joints_left = [2, 3, 4, 8, 9, 10]             joints_right = [5, 6, 7, 11, 12, 13]         else:             print('num_joints: %d' % num_joints)             raise KeyError("The dimension of adj matrix is wrong!")         adj_sym = torch.zeros_like(adj)         for i in range(num_joints):             for j in range(num_joints):                 if i == j:                     adj_sym[i][j] = 1                 if i in joints_left:                     index = joints_left.index(i)                     adj_sym[i][joints_right[index]] = 1.0                 if i in joints_right:                     index = joints_right.index(i)                     adj_sym[i][joints_left[index]] = 1.0         adj_1 = adj.matrix_power(1)         # store_2 = [3, 6, 10, 13, 16]         for i in np.arange(num_joints):             if i in store_2:                 adj_1[i] = 0         adj_2 = adj.matrix_power(2)         # store_2 = [3, 6, 10, 13, 16]         for i in np.arange(num_joints):             if i not in store_2:                 adj_2[i] = 0         adj_con = adj_1 + adj_2         self.gcn_sym = SemGraphConv(input_dim, output_dim, adj_sym)         self.bn_1 = nn.BatchNorm2d(output_dim, momentum=0.1)         self.gcn_con = SemGraphConv(input_dim, output_dim, adj_con)         self.bn_2 = nn.BatchNorm2d(output_dim, momentum=0.1)         self.relu = nn.ReLU()         self.cat_conv = nn.Conv2d(2 * output_dim, output_dim, 1, bias=False)         self.cat_bn = nn.BatchNorm2d(output_dim, momentum=0.1)         if dropout is not None:             self.dropout = nn.Dropout2d(dropout)         else:             self.dropout = None     def forward(self, input):         # x: (B, T, K, C)         x = self.gcn_sym(input)         y = self.gcn_con(input)         # x: (B, T, K, C) --&gt; (B, C, T, K)         x = x.permute(0, 3, 1, 2)         y = y.permute(0, 3, 1, 2)         x = self.relu(self.bn_1(x))         y = self.relu(self.bn_2(y))         output = torch.cat((x, y), dim=1)         output = self.cat_bn(self.cat_conv(output))         if self.dropout is not None:             output = self.dropout(self.relu(output))         else:             output = self.relu(output)         output = output.permute(0, 2, 3, 1)         return output
